### ISSUE TITLE
Type error for stream in parse-sam-header on sbcl 1.2.9 OSX 

### ISSUE THREAD

**Issue by keithj** [ExaScience/elprep#3]
I've got further from #2 by using the method I described. I've read the code and would expect this to work (via Gray streams), but nevertheless:

````
elPrep version 2.11. See http://github.com/exascience/elprep for more information.
Executing command:
  elprep /dev/stdin NA12878-chr22-10pct.only_mapped.bam --filter-unmapped-reads --sorting-order keep --gc-on 0 --nr-of-threads 1

debugger invoked on a TYPE-ERROR in thread
#<THREAD "main thread" RUNNING {10037AECC3}>:
  The value
    #S(ELPREP::BUFFERED-ASCII-INPUT-STREAM
       :INDEX 0
       :LIMIT 8192
       :BUFFER #(69 82 82 49 57 52 49 52 55 46 53 53 ...)
       :SECONDARY-BUFFER NIL
       :ELEMENT-TYPE BASE-CHAR
       :STREAM #<SB-SYS:FD-STREAM for "file /dev/fd/0" {10038C60C3}>)
  is not of type
    STREAM.

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

(no restarts: If you didn't do this on purpose, please report it as a bug.)

(ELPREP:PARSE-SAM-HEADER #<unavailable argument>) [tl,external]
0] 
````

**Comment by pcostanza** [comment]
Keith,

Thanks a lot for looking into elPrep. :)

Unfortunately, I cannot reproduce your issues (neither #2 nor #3). I'm on Mac OS X 10.10.2, and built sbcl 1.2.9 from scratch --with-sb-thread enabled, then built the elprep binary with the make-sbcl-binary.sh script. I'm using samtools 1.2 to read/write bam files. With that, executing elprep NA12878-chr22-10pct.bam out.bam --filter-unmapped-reads works fine.

Do you have a different setup? Can you post the exact elPrep command line you're invoking?

Thanks a lot,
Pascal

**Comment by keithj** [comment]
I'm on OSX 10.7.5. I used sbcl from Homebrew at first, but then wasn't confident that it had been built with all the desried features enabled. Then I built 1.2.9 from scratch to ensure that sb-thread was enabled.

I'm running the elprep-demo (small) script. I've found that the problem above is triggered by type declarations e.g.

````
(defun parse-sam-header (stream)
  "Parse a SAM file header section.
   See http://samtools.github.io/hts-specs/SAMv1.pdf - Section 1.3."
  (declare (stream stream) #.*optimization*)
  ...)

If I comment the type declaration then I can call functions in the REPL:

````
 (with-open-sam (s "toy.sam")
          (parse-sam-header s))

#S(SAM-HEADER
   :HD NIL
   :SQ ((:SN "ref" :LN 45) (:SN "ref2" :LN 40))
   :RG NIL
   :PG NIL
   :CO NIL
   :USER-TAGS NIL)
````

I will build sbcl again, carefully. I am having no problems with the Lispworks binary version.

**Comment by keithj** [comment]
````
*features*
(:NAMED-READTABLES :CL-FAD :BORDEAUX-THREADS :CL-PPCRE :THREAD-SUPPORT :SWANK
 :ASDF-PACKAGE-SYSTEM :ASDF3.1 :ASDF3 :ASDF2 :ASDF :OS-MACOSX :OS-UNIX
 :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :ALIEN-CALLBACKS :ANSI-CL
 :ASH-RIGHT-VOPS :BSD :C-STACK-IS-CONTROL-STACK :COMMON-LISP
 :COMPARE-AND-SWAP-VOPS :COMPLEX-FLOAT-VOPS :CYCLE-COUNTER :DARWIN
 :DARWIN9-OR-BETTER :FLOAT-EQL-VOPS :GENCGC :IEEE-FLOATING-POINT
 :INLINE-CONSTANTS :INODE64 :INTEGER-EQL-VOP :INTERLEAVED-RAW-SLOTS
 :LINKAGE-TABLE :LITTLE-ENDIAN :MACH-EXCEPTION-HANDLER :MACH-O
 :MEMORY-BARRIER-VOPS :MULTIPLY-HIGH-VOPS :OS-PROVIDES-BLKSIZE-T
 :OS-PROVIDES-DLADDR :OS-PROVIDES-DLOPEN :OS-PROVIDES-PUTWC
 :OS-PROVIDES-SUSECONDS-T :PACKAGE-LOCAL-NICKNAMES :PRECISE-ARG-COUNT-ERROR
 :RAW-INSTANCE-INIT-VOPS :SB-CORE-COMPRESSION :SB-DOC :SB-EVAL :SB-LDB
 :SB-PACKAGE-LOCKS :SB-SIMD-PACK :SB-SOURCE-LOCATIONS :SB-TEST :SB-THREAD
 :SB-UNICODE :SBCL :STACK-ALLOCATABLE-CLOSURES :STACK-ALLOCATABLE-FIXED-OBJECTS
 :STACK-ALLOCATABLE-LISTS :STACK-ALLOCATABLE-VECTORS
 :STACK-GROWS-DOWNWARD-NOT-UPWARD :SYMBOL-INFO-VOPS :UD2-BREAKPOINTS :UNIX
 :UNWIND-TO-FRAME-AND-CALL-VOP :X86-64)
````

**Comment by keithj** [comment]
I've rebuilt sbcl 1.2.9. This time I ran ````./make.sh --with-sb-thread```` rather than editing ````customize-target-features.lisp````. I did a hard reset of the git repository to remove all of my hacks and rebuilt elprep. However, both #2 and #3 persist.

**Comment by pcostanza** [comment]
This is very weird indeed. I double-checked everything on my end as well, and I really don't get these errors. However, the type declaration in parse-sam-header is effectively wrong, so we will remove it. Binding the standard streams instead of using setq should also be fine. So we will integrate those bug fixes. Thanks a lot for the reports!

Pascal

**Comment by keithj** [comment]
Thanks for your time. I'm sure the problem is somewhere in my OS X setup.

I repeated my build of SBCL 1.2.9. I rebuilt elprep, this time I deleted the ````$HOME/.cache/common-lisp```` directory between every attempt at building. ASDF is no longer doing what I expect when I modify a file and rebuild; it seems to be using a cache of old fasl files. Doing this before building the image gave me an executable which still exhibits #2, but not #3.

I'm intending to use the LispWorks delivered image on Linux as my main platform, which has been working perfectly.
